Documentation/diagnostics/source_link.md

```diff
@@ -0,0 +1,103 @@
+```json
+{
+    "docuemnts": {
+        "/usr/local/src/CodeFormatter/*": "https://raw.githubusercontent.com/dotnet/codeformatter/bcc51178e1a82fb2edaf47285f6e577989a7333f/*"
+    }
+}
+```
+
+Source Link configuration can also contain absolute mappings from file paths to URLs without an asterisk. Using the Code Formatter example from above, every source file could be mapped explicitly:
+```json
+{
+    "documents": {
+        "C:\\src\\CodeFormatter\\src\\CodeFormatter\\Program.cs":               "https://raw.githubusercontent.com/dotnet/codeformatter/bcc51178e1a82fb2edaf47285f6e577989a7333f/src/CodeFormatter/Program.cs",
+        "C:\\src\\CodeForamtter\\src\\CodeFormatter\\CommandLinerParser.cs":    "https://raw.githubusercontent.com/dotnet/codeformatter/bcc51178e1a82fb2edaf47285f6e577989a7333f/src/CodeFormatter/CommandLineParser.cs",
+        "C:\\src\\CodeFormatter\\src\\DeadRegions\\Program.cs":                 "https://raw.githubusercontent.com/dotnet/codeformatter/bcc51178e1a82fb2edaf47285f6e577989a7333f/src/DeadRegions/Program.cs",
+        "C:\\src\\CodeFormatter\\src\\DeadRegions\\Options.cs":                 "https://raw.githubusercontent.com/dotnet/codeformatter/bcc51178e1a82fb2edaf47285f6e577989a7333f/src/DeadRegions/Options.cs",
+        "C:\\src\\CodeFormatter\\src\\DeadRegions\\OptionParser.cs":            "https://raw.githubusercontent.com/dotnet/codeformatter/bcc51178e1a82fb2edaf47285f6e577989a7333f/src/DeadRegions/OptionParser.cs",
+        ...
+    }
+}
+```
+
+Source Link JSON may contain multiple relative and/or absolute mappings in any order. They will be resolved in order from most specific to least specific. Here is an example of a valid Source Link JSON for an arbitrary project structure:
+```json
+{
+    "C:\\src\\*":                   "http://MyDefaultDomain.com/src/*",
+    "C:\\src\\foo\\*":              "http://MyFooDomain.com/src/*",
+    "C:\\src\\foo\\specific.txt":   "http://MySpecificFoodDomain.com/src/specific.txt",
+    "C:\\src\\bar\\*":              "http://MyBarDomain.com/src/*"
+}
+```
+In this example:
+- All files under directory `bar` will map to a relative URL beginning with `http://MyBarDomain.com/src/`.
+- All files under directory `foo` will map to a relative URL beginning with `http://MyFooDomain.com/src/` **EXCEPT** `foo/specific.txt` which will map to `http://MySpecificFoodDomain.com/src/specific.txt`.
+- All other files anywhere under the `src` directory will map to a relative url beginning with `http://MyDefaultDomain.com/src/`.
+
+## Tooling
+- [Visual Studio 2017](https://www.visualstudio.com/downloads/)
+    - First debugger to support Source Link
+    - The current implementation uses a case insensitive string comparison between the file path and the Source Link entry. Using multiple entries that differ only in casing is not supported and the behavior is undefined. Here is an example of an unsupported Source Link JSON:
```

Comment:
The current implementation uses [qsort](http://www.cplusplus.com/reference/cstdlib/qsort/) and compares the strings based on length. qsort does not guarantee the ordering of equivalent elements.
> The order of equivalent elements is undefined.